proposals/NNNN-shader-execution-reordering.md

```diff
@@ -0,0 +1,1388 @@
+    bool enter = hit.GetHitKind() == HIT_KIND_TRIANGLE_FRONT_FACE;
+
+    payload.ior = UpdateIorList( newEntry, enter, iorList, iorListSize );
+
+    HitObject::Invoke( hit, payload );
+}
+```
+
+### Example: Do common computations with hit-coherence
+
+```C++
+hit = HitObject::TraceRay( ... );
+ReorderThread( hit );
+
+payload.giData = GlobalIlluminationCacheLookup( hit );
+
+HitObject::Invoke( hit, payload );
+```
+
+### Example: Same surface shader but different behavior in the raygeneration shader
+
+```C++
+// Primary ray wants perfect shadows and does not need to reorder as it is
+// coherent enough.
+ray = GeneratePrimaryRay();
+hit = HitObject::TraceRay( ... );
+RayDesc shadowRay = SampleShadow( hit );
+payload.shadowTerm = HitObject::TraceRay( shadowRay ).IsHit() ? 0.0f : 1.0f;
+HitObject::Invoke( hit, payload );
+
+// Secondary ray is incoherent but does not need perfect shadows.
+ray = ContinuePath( payload );
+hit = HitObject::TraceRay( ... );
+ReorderThread( hit );
+payload.shadowTerm = SampleShadowMap( hit );
+HitObject::Invoke( hit, payload );
+```
+
+### Example: Unified shading
+
+```C++
+hit = HitObject::TraceRay( ... );
+
+ReorderThread( hit );
+
+// Do not call HitObject::Invoke. Shade in raygeneration.
+```
+
+### Example: Coherently break render loop on miss
+
+Rationale: Executing the miss shader when not needed is unnecessarily
+inefficient on some architectures.
+
+```C++
+for( ;; )
+{
+    hit = HitObject::TraceRay( ... );
+    ReorderThread( hit );
+
+    if( hit.IsMiss() )
+        break;
+
+    HitObject::Invoke( hit, payload );
+}
+```
+
+### Example: Two-step shading, single reorder
+
+```C++
+hit = HitObject::TraceRay( ... );
+ReorderThread( hit );
+
+// Gather surface parameters into payload, e.g., compute normal and albedo
+// based on surface-specific functions and/or textures.
+HitObject::Invoke( hit, payload );
+
+// Alter the hit object to point to a unified surface shader.
+hit.SetShaderTableIndex( payload.surfaceShaderIdx );
+
+// Invoke unified surface shading. We are already hit-coherent so not worth
+// reordering again.
+HitObject::Invoke( hit, payload );
+```
+
+### Example: Live state optimization
+
+```C++
+hit = HitObject::TraceRay( ... );
+
+uint compressedNormal = CompressNormal( payload.normal ); 
+ReorderThread( hit );
+payload.normal = UncompressNormal( compressedNormal );
+
+HitObject::Invoke( hit, payload ); 
+```
+
+## DXIL specification
+
+The DXIL specification follows directly from the
+[detailed HLSL design](#detailed-design).
+The concept of a `HitObject` is represented by the DXIL type
+`dx.types.HitObject`:
+
+```DXIL
+%dx.types.HitObject = type { i8 * }
+```
+
+Central to `dx.types.HitObject` is its value semantics;
+an assignment is a copy.
+As the size of `dx.types.HitObject` is unknown at the DXIL level, the driver
+compiler will employ type replacement early in the lowering process. This is
+analogous to how `dx.types.Handle` is lowered. DXC will guarantee that the
```

Comment:
Under the circumstances, this document will have more longevity if we refer to it as the "offline compiler" or "Dxil generating compiler" rather than DXC.

Reply:
Fixed.